### PR TITLE
pkp/pkp-lib#8027 OPS preprints should support DOI versioning

### PR DESCRIPTION
**Pull request by ewhanson**

### CODE REVIEW

File: classes/doi/Repository.inc.php [hunk 1]
```diff
@@ -129,6 +129,43 @@ public function getDoisForSubmission(int $submissionId): array
...
+    /**
+     * Retrieves an application specific pubObject based on the type as a string
+     *
+     * @param string $pubObjectType The object type name passed down from the front end. Will be application specific, e.g. 'article' or 'preprint' instead of 'publication'
+     *
+     * @return DataObject|null Will be one of the application specific pubObjects, e.g. Publication, ArticleGalley, etc.
+     */
+    public function getPubObjectFromType(string $pubObjectType, int $pubObjectId): ?DataObject
+    {
+        return match ($pubObjectType) {
```
[the remark this answers is not in the document]
Comment (asmecher) [reply classes/doi/Repository.inc.php:141]: :+1: I've been unreasonably excited about `match` and am looking forward to replacing some really ugly formulations with it!

File: classes/components/forms/context/DoiSetupSettingsForm.php [hunk 2]
```diff
@@ -40,6 +41,22 @@ public function __construct(string $action, array $locales, Context $context)
...
+                'description' => __('doi.manager.settings.doiVersioning.description'),
+                'groupId' => self::DOI_SETTINGS_GROUP,
+                'options' => [
+                    [
+                        'value' => true,
+                        'label' => __('common.yes')
+                    ],
+                    [
+                        'value' => false,
+                        'label' => __('common.no')
```
Comment (NateWr) [inline classes/components/forms/context/DoiSetupSettingsForm.php:56]: I prefer option labels that offer more confirmation about what is being selected, like:

> - [ ] Yes, assign a unique DOI to every version of a preprint.
> - [ ] No, all versions of a preprint should have the same DOI.

File: schemas/context.json [hunk 3]
```diff
@@ -89,6 +89,10 @@
 			"validation": [
 				"nullable"
 			]
+		},
+		"doiVersioning": {
+			"type": "boolean",
+			"default": true
```
Comment (NateWr) [inline schemas/context.json:95]: Can you add a "description" to this that says what it does? (This filters through to the REST API docs.)

**Review by NateWr — CHANGES_REQUESTED** [review]

File: api/v1/_dois/BackendDoiHandler.php [hunk 4]
```diff
@@ -54,10 +54,6 @@ public function editGalley(SlimRequest $slimRequest, APIResponse $response, arra
             return $response->withStatus(404)->withJsonError('api.404.resourceNotFound');
         }
 
-        if ($galley->getData('contextId') !== $context->getId()) {
-            return $response->withStatus(403)->withJsonError('api.dois.403.editItemOutOfContext');
-        }
-
```
Comment (bozana) [inline api/v1/_dois/BackendDoiHandler.php:60]: Here too: Maybe to get the submission here already and check if it belongs to the context?

**Review by bozana — COMMENTED** [review]
I haven't taken a deeper look at the test. Will do it as soon as possible...

File: templates/frontend/objects/preprint_details.tpl [hunk 5]
```diff
@@ -149,7 +149,12 @@
 			{/if}
 
 			{* DOI *}
-			{assign var=doiObject value=$currentPublication->getData('doiObject')}
+			{if $currentContext->getData('doiVersioning') == true}
+				{assign var=doiObject value=$publication->getData('doiObject')}
+			{else}
+				{assign var=doiObject value=$currentPublication->getData('doiObject')}
+			{/if}
+
```
Comment (bozana) [inline templates/frontend/objects/preprint_details.tpl:157]: Maybe we do not need this check now that we save DOIs for all publications even if the versioning is off?

Comment (ewhanson) [reply templates/frontend/objects/preprint_details.tpl:157]: Good idea. That way we can get it directly from the publication without worrying about whether it's the current publication or not.

**Pull request merged**